src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java

```diff
@@ -155,6 +186,71 @@ private Mono<Void> validateOpaqueReferenceToken(String issBaseUri, String token,
                 });
     }
 
+    /**
+     * Validates a token's audience or client ID depending on token type.
+     *
+     * JWT ID tokens (representing end users) typically contain an 'aud' claim that needs validation.
+     * JWT access tokens may not have an 'aud' claim but instead use a 'client_id' claim.
+     *
+     * We first try to validate the audience. Only if no valid audience is found, we fall back to
+     * client ID validation. This allows both token types to work with the gateway.
+     *
+     * IMPORTANT NOTES:
+     * - Currently, we only allow GridSuite audience tokens in the allowedAudiences configuration
+     * - If we want to allow other frontend applications to access this API:
+     *   a) This validation logic needs to be reviewed and possibly expanded
+     *   b) The CORS strategy would need to be modified accordingly to allow those origins
+     *
+     * @param jwtClaimsSet The JWT claims set
+     * @param exchange The server web exchange
+     * @return Mono<Void> with unauthorized response if validation fails, null if validation passes
+     */
+    private Mono<Void> validateAudienceOrClientId(JWTClaimsSet jwtClaimsSet, ServerWebExchange exchange) {
+        LOGGER.debug("checking audience or client ID");
+        List<String> tokenAudiences = jwtClaimsSet.getAudience();
+
+        boolean audienceMatched = false;
+        if (tokenAudiences != null && !tokenAudiences.isEmpty()) {
+            audienceMatched = tokenAudiences.stream()
+                    .anyMatch(aud -> allowedAudiences.contains(aud));
+        }
+
+        if (audienceMatched) {
+            LOGGER.debug("Audience validation successful");
+            return null;
+        } else {
+            LOGGER.debug("Audience validation failed for audiences: {}, trying client ID as fallback", tokenAudiences);
+            String clientIdClaim = (String) jwtClaimsSet.getClaim("client_id");
+            if (clientIdClaim != null && allowedClients.contains(clientIdClaim)) {
+                LOGGER.debug("Client ID validation successful");
+                return null;
+            }
+        }
+
+        if (tokenAudiences != null && !tokenAudiences.isEmpty()) {
+            LOGGER.info(UNAUTHORIZED_AUDIENCE_NOT_ALLOWED, exchange.getRequest().getPath(), tokenAudiences);
+        } else {
+            LOGGER.info(UNAUTHORIZED_CLIENT_NOT_ALLOWED, exchange.getRequest().getPath(), jwtClaimsSet.getClaim("client_id"));
+        }
+
+        return completeWithCode(exchange, HttpStatus.UNAUTHORIZED);
+    }
+
+    /**
+     * Validates whether a client ID is in the list of allowed clients.
+     *
+     * @param clientId The client ID to validate
+     * @param exchange The server web exchange
+     * @return Mono<Void> with unauthorized response if validation fails, null if validation passes
+     */
+    private Mono<Void> validateClientId(String clientId, ServerWebExchange exchange) {
```

Comment:
Same Maybe better to return a synchronous value (boolean true false, or void/exception like the lib validator) and completeWithCode in the caller